File: project_euler/problem_302/solution.py
"""
Problem 302: Strong Achilles Numbers below 10^18.

A Strong Achilles number n satisfies:
  (1) n is powerful (all prime exponents >= 2),
  (2) n is not a perfect power (gcd of exponents == 1),
  (3) phi(n) is also Achilles.

Enumerate powerful numbers via DFS over prime factorizations,
checking all four conditions at each node.
"""

from math import gcd
from functools import reduce
from sympy import factorint, primerange

LIMIT = 10**18
small_primes = list(primerange(2, 10**6 + 1))


def is_achilles(fac):
    """Check if a number with factorization {p: e} is Achilles."""
    if not fac:
        return False
    exps = list(fac.values())
    if any(e < 2 for e in exps):
        return False
    return reduce(gcd, exps) == 1


def phi_factorization(prime_exp_list):
    """Compute the factorization of phi(n) given n = prod p_i^{e_i}."""
    phi_fac = {}
    for p, e in prime_exp_list:
        if e - 1 > 0:
            phi_fac[p] = phi_fac.get(p, 0) + (e - 1)
        for q, f in factorint(p - 1).items():
            phi_fac[q] = phi_fac.get(q, 0) + f
    return phi_fac


answer = 0


def dfs(idx, current_n, exponents, gcd_exp):
    global answer
    if exponents:
        if gcd_exp == 1:
            phi_fac = phi_factorization(exponents)
            if is_achilles(phi_fac):
                answer += 1

    for i in range(idx, len(small_primes)):
        p = small_primes[i]
        pp = p * p
        if current_n * pp > LIMIT:
            break
        power = pp
        e = 2
        while current_n * power <= LIMIT:
            new_gcd = e if gcd_exp == 0 else gcd(gcd_exp, e)
            exponents.append((p, e))
            dfs(i + 1, current_n * power, exponents, new_gcd)
            exponents.pop()
            if power > LIMIT // p:
                break
            power *= p
            e += 1


dfs(0, 1, [], 0)
print(answer)

File: project_euler/problem_302/solution.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Problem 302: Strong Achilles Numbers below 10^18.
 *
 * Enumerate powerful numbers n < 10^18 via DFS over prime factorizations.
 * Check: (1) all exponents >= 2, (2) gcd of exponents = 1 (Achilles),
 * (3)-(4) phi(n) is also Achilles.
 */

typedef long long ll;
typedef __int128 lll;
const ll LIMIT = 1000000000000000000LL;

vector<int> primes;

void gen_primes(int lim) {
    vector<bool> sieve(lim + 1, false);
    for (int i = 2; i <= lim; i++) {
        if (!sieve[i]) {
            primes.push_back(i);
            for (ll j = (ll)i * i; j <= lim; j += i)
                sieve[j] = true;
        }
    }
}

vector<pair<int, int>> factorize(ll n) {
    vector<pair<int, int>> res;
    for (int p : primes) {
        if ((ll)p * p > n) break;
        if (n % p == 0) {
            int e = 0;
            while (n % p == 0) { n /= p; e++; }
            res.push_back({p, e});
        }
    }
    if (n > 1) res.push_back({(int)n, 1});
    return res;
}

unordered_map<int, vector<pair<int,int>>> pm1_cache;

const vector<pair<int,int>>& get_pm1_fac(int p) {
    auto it = pm1_cache.find(p);
    if (it != pm1_cache.end()) return it->second;
    pm1_cache[p] = factorize(p - 1);
    return pm1_cache[p];
}

ll answer = 0;

void dfs(int idx, lll current_n, int num_primes, int gcd_exp,
         unordered_map<int, int>& phi_fac) {
    if (num_primes >= 2 && gcd_exp == 1) {
        bool powerful = true;
        int phi_gcd = 0;
        for (auto& [p, e] : phi_fac) {
            if (e < 2) { powerful = false; break; }
            phi_gcd = __gcd(phi_gcd, e);
        }
        if (powerful && phi_gcd == 1)
            answer++;
    }

    for (int i = idx; i < (int)primes.size(); i++) {
        ll p = primes[i];
        if (current_n * (lll)p * p > LIMIT) break;

        const auto& pm1_fac = get_pm1_fac(p);
        lll power = (lll)p * p;
        for (int e = 2; current_n * power <= LIMIT; e++) {
            int new_gcd = (gcd_exp == 0) ? e : __gcd(gcd_exp, e);
            phi_fac[p] += (e - 1);
            for (auto& [q, f] : pm1_fac) phi_fac[q] += f;

            dfs(i + 1, current_n * power, num_primes + 1, new_gcd, phi_fac);

            for (auto& [q, f] : pm1_fac) {
                phi_fac[q] -= f;
                if (phi_fac[q] == 0) phi_fac.erase(q);
            }
            phi_fac[p] -= (e - 1);
            if (phi_fac[p] == 0) phi_fac.erase(p);

            if (power > (lll)LIMIT / p) break;
            power *= p;
        }
    }
}

int main() {
    gen_primes(1000000);
    unordered_map<int, int> phi_fac;
    dfs(0, 1, 0, 0, phi_fac);
    printf("%lld\n", answer);
    return 0;
}
